src/util/heap/layout/fragmented_mapper.rs

```diff
@@ -107,6 +107,7 @@ impl Mmapper for FragmentedMapper {
                         let mmap_start = Self::chunk_index_to_address(base, chunk);
                         let _guard = self.lock.lock().unwrap();
                         crate::util::memory::dzmmap(mmap_start, MMAP_CHUNK_BYTES).unwrap();
+                        notify_mmap(mmap_start);
```

Comment:
Call the callback function only when the chunk is not mapped before.

Reply:
I suggest adding two arguments to `Mmapper::ensure_mapped` e.g. change it to `Mmapper::ensure_mapped(start, pages, global_metadata_per_chunk: usize, local_metadata_per_chunk: usize)`, and probably do the same for `dzmmap()`.

Reply:
This will be slightly different if we switch from per-chunk side metadata, but is conceptually the same

Reply:
Why is this callback only used in `FragmentedMapper` (64-bit only)? You removed the side metadata mmap code from `Space.init()` which is used for both 32bit and 64bit. Does this mean we are not mmapping data for contiguous spaces in 32 bits?

Reply:
One design problem of this fix is, `Space` is the caller of this mmapper code. When a space is notified by a mmaper (using `notify_mmap`), it needs to map both space-specific and global metadata. But spaces shouldn't access/manage global metadata by design.

Reply:
Anyway, we still need to refactor this mmap code at some point in the future.  When a mmtk heap chunk is mapped by some third-party other than mmtk, we should not do metadata memory mapping.

Reply:
> I suggest adding two arguments to `Mmapper::ensure_mapped` e.g. change it to `Mmapper::ensure_mapped(start, pages, global_metadata_per_chunk: usize, local_metadata_per_chunk: usize)`, and probably do the same for `dzmmap()`.

Fixed